src/main/java/com/hierynomus/smbj/transport/BaseTransport.java

```diff
@@ -57,5 +61,39 @@ public void write(SMB2Packet packet) throws TransportException {
         }
     }
 
+    @Override
+    public void writeSigned(SMB2Packet packet, byte[] signingKey) throws TransportException {
+        writeLock.lock();
+        try {
+            try {
+                SMBBuffer buffer = new SMBBuffer();
+                packet.getHeader().setFlag(SMB2MessageFlag.SMB2_FLAGS_SIGNED); // set the SMB2_FLAGS_SIGNED flag
+                packet.write(buffer);
+                
+                signBuffer(buffer, signingKey);
+                
+                logger.trace("Writing packet << {} >>, sequence number << {} >>", packet.getHeader().getMessage(), packet.getSequenceNumber());
+                doWrite(buffer);
+                out.flush();
+            } catch (IOException ioe) {
+                throw new TransportException(ioe);
+            } catch (InvalidKeyException e) {
+                throw new TransportException(e);
+            } catch (NoSuchAlgorithmException e) {
+                throw new TransportException(e);
+            }
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
     protected abstract void doWrite(SMBBuffer packetData) throws IOException;
+
+    
+    // [MS-SMB2] 3.1.4.1 Signing An Outgoing Message
+    // If Connection.Dialect is "2.0.2" or "2.1", the sender MUST compute a 32-byte hash using HMAC-SHA256 over the entire message, 
```

Comment:
This is my first foray into your codebase, so I didn't want to make assumptions about how you'd like to design that.